### ISSUE TITLE
dialyzer warns for sign_in/4

### ISSUE THREAD

**Issue by hykw** [ueberauth/guardian#252]
When I use dialyzer, it warns like the following.

```
web/controllers/page_controller.ex:4: Function index/2 has no local return
web/controllers/page_controller.ex:14: The call 'Elixir.Guardian.Plug':sign_in(conn@1::#{},user@1::#{},'token',[]) breaks the contract ('Elixir.Plug.Conn':t(),any(),atom() | 'Elixir.String':t(),#{}) -> 'Elixir.Plug.Conn':t()
```

The functions, I confirmed, it cause the warning are these(both of them are warned).

```elixir
Guardian.Plug.sign_in(conn, user, :token, [])
Guardian.Plug.sign_in(conn, user, :token, perms: %{user: [:normal_login]})
```

I confirmed it in guardian 0.14.2 and dialyxir 0.4.3.
The test code is here: https://github.com/hykw/guardian_dialyzer_specimen

In the following method, sign_in/3 and sign_in/4(the forth param is just %{}) emits no warning, but in case the last one is List or Keyword, it would be warned.

```elixir
defmodule Specimen.PageController do
  use Specimen.Web, :controller

  def index(conn, _params) do
    user = %Specimen.User{}
    perms = %{user: [:normal_login]}

    # OK
    Guardian.Plug.sign_in(conn, user, :token)
    Guardian.Plug.sign_in(conn, user, :token, %{})


    # NG
    Guardian.Plug.sign_in(conn, user, :token, [])
    Guardian.Plug.sign_in(conn, user, :token, perms: perms)

    render conn, "index.html"
  end

end
```

**Comment by hykw** [comment]
The warning I pasted was run in Erlang 18.2.1
In an Erlang 19.2 environment, the warning differed like below

```
web/controllers/page_controller.ex:4: Function index/2 has no local return
web/controllers/page_controller.ex:14: The call 'Elixir.Guardian.Plug':sign_in(conn@1::#{'__struct__':='Elixir.Plug.Conn', 'adapter':={atom(),_}, 'assigns':=#{atom()=>_}, 'before_send':=[fun((#{'__struct__':='Elixir.Plug.Conn', 'adapter':={_,_}, 'assigns':=map(), 'before_send':=[fun((_) -> any())], _=>_}) -> #{'__struct__':='Elixir.Plug.Conn', 'adapter':={_,_}, 'assigns':=map(), 'before_send':=[fun((_) -> any())], _=>_})], 'body_params':=#{'__struct__'=>'Elixir.Plug.Conn.Unfetched', 'aspect'=>atom(), binary()=>binary() | [binary() | [any()] | #{binary()=>_}] | #{binary()=>binary() | [any()] | #{binary()=>_}}}, 'cookies':=#{'__struct__'=>'Elixir.Plug.Conn.Unfetched', 'aspect'=>atom(), binary()=>binary()}, 'halted':=_, 'host':=binary(), 'method':=binary(), 'owner':=pid(), 'params':=#{'__struct__'=>'Elixir.Plug.Conn.Unfetched', 'aspect'=>atom(), binary()=>binary() | [binary() | [any()] | #{binary()=>_}] | #{binary()=>binary() | [any()] | #{binary()=>_}}}, 'path_info':=[binary()], 'path_params':=#{binary()=>binary() | [binary() | [any()] | #{binary()=>_}] | #{binary()=>binary() | [any()] | #{binary()=>_}}}, 'peer':={{byte(),byte(),byte(),byte()} | {char(),char(),char(),char(),char(),char(),char(),char()},char()}, 'port':=char(), 'private':=#{atom()=>_}, 'query_params':=#{'__struct__'=>'Elixir.Plug.Conn.Unfetched', 'aspect'=>atom(), binary()=>binary() | [binary() | [any()] | #{binary()=>_}] | #{binary()=>binary() | [any()] | #{binary()=>_}}}, 'query_string':=binary(), 'remote_ip':={byte(),byte(),byte(),byte()} | {char(),char(),char(),char(),char(),char(),char(),char()}, 'req_cookies':=#{'__struct__'=>'Elixir.Plug.Conn.Unfetched', 'aspect'=>atom(), binary()=>binary()}, 'req_headers':=[{binary(),binary()}], 'request_path':=binary(), 'resp_body':='nil' | binary() | maybe_improper_list(binary() | maybe_improper_list(any(),binary() | []) | byte(),binary() | []), 'resp_cookies':=#{binary()=>#{}}, 'resp_headers':=[{binary(),binary()}], 'scheme':='http' | 'https', 'script_name':=[binary()], 'secret_key_base':='nil' | binary(), 'state':='chunked' | 'file' | 'sent' | 'set' | 'unset', 'status':='nil' | non_neg_integer()},user@1::#{'__struct__':='Elixir.Specimen.User', 'id':=1},'token',[]) breaks the contract ('Elixir.Plug.Conn':t(),any(),atom() | 'Elixir.String':t(),map()) -> 'Elixir.Plug.Conn':t()
```

**Comment by doomspork** [comment]
Thanks for the report @hykw, I'll take a look 👍

**Comment by scrogson** [comment]
This appears to have been fixed in #256
